piq/iw_ssim.py

```diff
@@ -0,0 +1,458 @@
+                              kernel_size: int = 11, kernel_sigma: float = 1.5, k1: float = 0.01, k2: float = 0.03,
+                              parent: bool = True, blk_size: int = 3, sigma_nsq: float = 0.4,
+                              scale_weights: Optional[torch.Tensor] = None,
+                              reduction: str = 'mean') -> torch.Tensor:
+    r"""Interface of Information Content Weighted Structural Similarity (IW-SSIM) index.
+    Inputs supposed to be in range ``[0, data_range]``.
+
+    Args:
+        x: An input tensor. Shape :math:`(N, C, H, W)`.
+        y: A target tensor. Shape :math:`(N, C, H, W)`.
+        data_range: Maximum value range of images (usually 1.0 or 255).
+        kernel_size: The side-length of the sliding window used in comparison. Must be an odd value.
+        kernel_sigma: Sigma of normal distribution for sliding window used in comparison.
+        k1: Algorithm parameter, K1 (small constant).
+        k2: Algorithm parameter, K2 (small constant).
+            Try a larger K2 constant (e.g. 0.4) if you get a negative or NaN results.
+        parent: Flag to control dependency on previous layer of pyramid.
+        blk_size: The side-length of the sliding window used in comparison for information content.
+        sigma_nsq: Parameter of visual distortion model.
+        scale_weights: Weights for scaling.
+        reduction: Specifies the reduction type:
+            ``'none'`` | ``'mean'`` | ``'sum'``. Default:``'mean'``
+
+    Returns:
+        Value of Information Content Weighted Structural Similarity (IW-SSIM) index.
+
+    References:
+        Wang, Zhou, and Qiang Li..
+        Information content weighting for perceptual image quality assessment.
+        IEEE Transactions on image processing 20.5 (2011): 1185-1198.
+        https://ece.uwaterloo.ca/~z70wang/publications/IWSSIM.pdf DOI:`10.1109/TIP.2010.2092435`
+
+    Note:
+        Lack of content in the image could lead to NaN values.
+    """
+    assert kernel_size % 2 == 1, f'Kernel size must be odd, got [{kernel_size}]'
+
+    _validate_input(tensors=[x, y], dim_range=(4, 4), data_range=(0., data_range))
+
+    x = x / float(data_range) * 255
+    y = y / float(data_range) * 255
+
+    if x.size(1) == 3:
+        x = rgb2yiq(x)[:, :1]
+        y = rgb2yiq(y)[:, :1]
+
+    if scale_weights is None:
+        scale_weights = torch.tensor([0.0448, 0.2856, 0.3001, 0.2363, 0.1333], dtype=x.dtype, device=x.device)
+    scale_weights = scale_weights / scale_weights.sum()
+    assert scale_weights.size(0) == scale_weights.numel(),\
+        f'Expected a vector of weights, got {scale_weights.dim()}D tensor'
+
+    levels = scale_weights.size(0)
+
+    min_size = (kernel_size - 1) * 2 ** (levels - 1) + 1
+    if x.size(-1) < min_size or x.size(-2) < min_size:
+        raise ValueError(f'Invalid size of the input images, expected at least {min_size}x{min_size}.')
```

Comment:
You test here that input satisfies requirements. 5 lines above when testing that scale weights satisfy requirements you raise an assertion, but here you raise exception. Let's be consistent.
I vote for raising assertion, because they can be switched off and improve performance.